x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java

```diff
@@ -1249,60 +1255,137 @@ public String getFeatureDescription() {
 
+        ActionListener<ListTasksResponse> afterWaitingForTasks = ActionListener.wrap(
+            listTasksResponse -> {
+                listTasksResponse.rethrowFailures("Waiting for indexing requests for .ml-* indices");
+                if (results.values().stream().allMatch(b -> b)) {
+                    // Call into the original listener to clean up the indices
+                    SystemIndexPlugin.super.cleanUpFeature(clusterService, client, unsetResetModeListener);
+                } else {
+                    final List<String> failedComponents = results.entrySet().stream()
+                        .filter(result -> result.getValue() == false)
+                        .map(Map.Entry::getKey)
+                        .collect(Collectors.toList());
+                    unsetResetModeListener.onFailure(new RuntimeException("Some components failed to reset: " + failedComponents));
+                }
+            },
+            unsetResetModeListener::onFailure
+        );
+
         ActionListener<StopDataFrameAnalyticsAction.Response> afterDataframesStopped = ActionListener.wrap(dataFrameStopResponse -> {
             // Handle the response
             results.put("data_frame/analytics", dataFrameStopResponse.isStopped());
-
             if (results.values().stream().allMatch(b -> b)) {
-                // Call into the original listener to clean up the indices
-                SystemIndexPlugin.super.cleanUpFeature(clusterService, client, listener);
+                client.admin()
+                    .cluster()
+                    .prepareListTasks()
+                    .setActions("xpack/ml/*")
+                    .setWaitForCompletion(true)
+                    .execute(ActionListener.wrap(
+                        listMlTasks -> {
+                            listMlTasks.rethrowFailures("Waiting for machine learning tasks");
+                            client.admin()
+                                .cluster()
+                                .prepareListTasks()
+                                .setActions("indices:data/write/bulk")
+                                .setDetailed(true)
+                                .setWaitForCompletion(true)
+                                .setDescriptions("*.ml-*")
+                                .execute(afterWaitingForTasks);
+                        },
+                        unsetResetModeListener::onFailure
+                    ));
             } else {
                 final List<String> failedComponents = results.entrySet().stream()
                     .filter(result -> result.getValue() == false)
                     .map(Map.Entry::getKey)
                     .collect(Collectors.toList());
-                listener.onFailure(new RuntimeException("Some components failed to reset: " + failedComponents));
+                unsetResetModeListener.onFailure(new RuntimeException("Some components failed to reset: " + failedComponents));
```

Comment:
```suggestion
                unsetResetModeListener.onFailure(new RuntimeException("Some machine learning components failed to reset: " + failedComponents));
```

Since the user could be resetting multiple features simultaneously, I think all the error responses should make clear they relate to ML.